src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1050,11 +1051,32 @@ template <>
 Value emitScalarOpFor<ONNXEqualOp>(ConversionPatternRewriter &rewriter,
     Location loc, Operation *op, Type elementType,
     ArrayRef<Value> scalarOperands) {
+
   CheckIfCustomScalarOpIsSupported<ONNXEqualOp>(elementType);
+  Value results;
   Value lhs = scalarOperands[0];
   Value rhs = scalarOperands[1];
-  MultiDialectBuilder<MathBuilder> create(rewriter, loc);
-  return create.math.eq(lhs, rhs);
+  MultiDialectBuilder<KrnlBuilder, MathBuilder> create(rewriter, loc);
+  Type inputElemType = getElementType(lhs.getType());
+
+  // If the two input values are a string then we want to use the krnlstrncmp.
+  // However, if the input values are a float or an int we can simply use the
+  // equal function.
+  if (inputElemType.isa<krnl::StringType>()) {
+    Value strlenRes = create.krnl.strlen(lhs);
+    Value strncmpRes = create.krnl.strncmp(lhs, rhs, strlenRes);
+    // We need to convert the results to return *i1 since krnlstrncmp returns
+    // i32.
+    Value convertedRes = create.math.cast(elementType, strncmpRes);
```

Comment:
I mean this cast is redundant. It's enough to have:
```c
 Value zeroVal = create.math.constant(convertedRes.getType(), 0);
 results = create.math.eq(convertedRes, zeroVal);
```

Reply:
Okay I can remove it.... you said `this looks redundant, we can create a zeroVal with strncmpRes.getType() then compare it to `convertedRes'.` and convertedRes is the cast so I was confused, I think that is why I put we do not need it.

Reply:
I removed it.

Reply:
Thanks! and sorry for the confusion.